File: cable_joints/commonSystems.js
import Vector2 from './vector2.js';

import {
  PositionComponent,
  VelocityComponent,
  GravityAffectedComponent,
  BallTagComponent,
  RadiusComponent,
  MassComponent,
  RestitutionComponent,
  PrevFinalPosComponent,
  OrientationComponent,
  MomentOfInertiaComponent,
  AngularVelocityComponent,
  CoefficientOfFrictionComponent,
  ObstacleTagComponent,
  ObstaclePushComponent,
  ScoredTagComponent,
  DistanceConstraintComponent,
  PrevFinalOrientationComponent,
  FlipperTagComponent,
  FlipperTipComponent,
  FlipperStateComponent,
  BorderComponent
} from './ecs.js';

import {
  closestPointOnSegment
} from './geometry.js';


export class PrevFinalOrientationSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([OrientationComponent, PrevFinalOrientationComponent]);
        for (const entityId of entities) {
            const orientationComp = world.getComponent(entityId, OrientationComponent);
            const prevFinalOrientationComp = world.getComponent(entityId, PrevFinalOrientationComponent);
            prevFinalOrientationComp.angle = orientationComp.angle;
        }
    }
}

export class PBDAngularVelocityUpdateSystem {
  runInPause = false;

  _normalizeAngle(angle) { // Helper to find the shortest angle difference
    while (angle > Math.PI) angle -= 2 * Math.PI;
    while (angle < -Math.PI) angle += 2 * Math.PI;
    return angle;
  }

  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const entities = world.query([
      OrientationComponent,
      AngularVelocityComponent,
      PrevFinalOrientationComponent,
      MomentOfInertiaComponent
    ]);
    const epsilon = 1e-9;

    if (dt <= epsilon) return;

    for (const entityId of entities) {
      if (entityId === grabbed) continue;

      const moiComp = world.getComponent(entityId, MomentOfInertiaComponent);
      if (moiComp && moiComp.invInertia <= 0) continue; // Skip static or non-rotational objects

      const orientationComp = world.getComponent(entityId, OrientationComponent);
      const angularVelComp = world.getComponent(entityId, AngularVelocityComponent);
      const prevFinalOrientationComp = world.getComponent(entityId, PrevFinalOrientationComponent);

      const currentAngle = orientationComp.angle;
      const prevAngle = prevFinalOrientationComp.angle;

      let deltaAngle = this._normalizeAngle(currentAngle - prevAngle);

      angularVelComp.angularVelocity = deltaAngle / dt;
    }
  }
}

export class PBDVelocityUpdateSystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const entities = world.query([PositionComponent, VelocityComponent, PrevFinalPosComponent, MassComponent]);
    const epsilon = 1e-9;

    if (dt <= epsilon) return; // Avoid division by zero or very small dt

    for (const entityId of entities) {
      if (entityId === grabbed) continue; // Don't update velocity of grabbed ball this way

      const massComp = world.getComponent(entityId, MassComponent);
      // Only update velocities for dynamic objects (mass > 0)
      if (massComp && massComp.mass <= 0) continue;


      const posComp = world.getComponent(entityId, PositionComponent);
      const velComp = world.getComponent(entityId, VelocityComponent);
      const prevFinalPosComp = world.getComponent(entityId, PrevFinalPosComponent);

      // v = (x_new - x_old) / dt
      velComp.vel.subtractVectors(posComp.pos, prevFinalPosComp.pos).scale(1.0 / dt);
    }
  }
}


export class FlipperTipLinkSystem {
  runInPause = false;
  update(world, dt) {
    // for each tip‐entity, compute its current position
    for (const tipId of world.query([FlipperTipComponent, PositionComponent, FlipperTipComponent])) {
      const tipComp   = world.getComponent(tipId, FlipperTipComponent);
      const flipId    = tipComp.flipperEntityId;
      const pivotPos  = world.getComponent(flipId, PositionComponent).pos;
      const state     = world.getComponent(flipId, FlipperStateComponent);
      // same math you use elsewhere to find the tip
      const angle = state.restAngle + state.sign * state.rotation;
      const dir   = new Vector2(Math.cos(angle), Math.sin(angle));
      const tipPos = pivotPos.clone().add(dir, state.length);
      // write it into the tip entity’s PositionComponent
      world.getComponent(tipId, PositionComponent).pos.set(tipPos);
    }
  }
}

export class GravitySystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    const gravity = world.getResource('gravity');
    if (!gravity) return;

    const entities = world.query([VelocityComponent, GravityAffectedComponent]);
    for (const entityId of entities) {
      if (entityId === grabbed) continue;
      const velComp = world.getComponent(entityId, VelocityComponent);
      velComp.vel.add(gravity, dt);
    }
  }
}

export class XPBDDistanceConstraintSystem {
  runInPause = false;
  update(world, dt) {
    const constraintEntities = world.query([DistanceConstraintComponent]);
    const epsilon = 1e-9; // Small value to avoid division by zero

    for (const entityId of constraintEntities) {
      const constraint = world.getComponent(entityId, DistanceConstraintComponent);

      const entityA = constraint.entityA;
      const entityB = constraint.entityB;

      const posAComp = world.getComponent(entityA, PositionComponent);
      const posBComp = world.getComponent(entityB, PositionComponent);

      if (!posAComp || !posBComp) continue;

      const pA = posAComp.pos;
      const pB = posBComp.pos;

      const massAComp = world.getComponent(entityA, MassComponent);
      const invMassA = (massAComp && massAComp.mass > 0) ? 1.0 / massAComp.mass : 0.0;
      const moiAComp = world.getComponent(entityA, MomentOfInertiaComponent);
      const invInertiaA = moiAComp ? moiAComp.invInertia : 0.0;

      const massBComp = world.getComponent(entityB, MassComponent);
      const invMassB = (massBComp && massBComp.mass > 0) ? 1.0 / massBComp.mass : 0.0;
      const moiBComp = world.getComponent(entityB, MomentOfInertiaComponent);
      const invInertiaB = moiBComp ? moiBComp.invInertia : 0.0;

      // If both entities are effectively immovable, skip
      if (invMassA + invMassB + invInertiaA + invInertiaB <= epsilon) {
        continue;
      }

      const diff = new Vector2().subtractVectors(pB, pA);
      const currentLength = diff.length();
      if (currentLength <= epsilon) continue;

      const dir = diff.clone().scale(1.0 / currentLength); // Normalized direction from A to B

      const C = currentLength - constraint.restLength; // Constraint violation

      // XPBD compliance term (alpha_tilde in the paper)
      const alpha_tilde = constraint.compliance / (dt * dt);

      // Generalized inverse mass for the constraint
      // For distance constraints, the gradient w.r.t position is 'dir' or '-dir'
      // For rotation, it's more complex if attachment points are offset, but here we assume CoM constraint
      // For simplicity, assuming constraint acts on CoM directly (no rotational coupling here)
      // More general XPBD would include r x n terms for rotational parts.
      // This simplified version matches standard PBD distance constraint generalized mass.
      let w_sum = invMassA + invMassB; // Simplified: only translational

      // Denominator for delta_lambda
      const denominator = w_sum + alpha_tilde;
      if (denominator <= epsilon) continue;

      const delta_lambda = (-C - alpha_tilde * constraint.lambda) / denominator;
      constraint.lambda += delta_lambda;

      const correction = dir.clone().scale(delta_lambda);

      // Apply positional corrections
      if (invMassA > 0) {
        const dpA = correction.clone().scale(-invMassA);
        pA.add(dpA);
        const velAComp = world.getComponent(entityA, VelocityComponent);
        if (velAComp && dt > epsilon) {
            // velAComp.vel.add(dpA, 1.0 / dt); // This would be for PBD velocity update
        }
      }
      if (invMassB > 0) {
        const dpB = correction.clone().scale(invMassB);
        pB.add(dpB);
        const velBComp = world.getComponent(entityB, VelocityComponent);
        if (velBComp && dt > epsilon) {
            // velBComp.vel.add(dpB, 1.0 / dt); // This would be for PBD velocity update
        }
      }
    }
  }
}

export class MovementSystem {
  runInPause = false;
  update(world, dt) {
    const grabbed = world.getResource('grabbedBall');
    // Update linear position
    const linearEntities = world.query([PositionComponent, VelocityComponent]);
    for (const entityId of linearEntities) {
      if (entityId === grabbed) continue;
      const posComp = world.getComponent(entityId, PositionComponent);
      const velComp = world.getComponent(entityId, VelocityComponent);
      posComp.pos.add(velComp.vel, dt);
    }
  }
}

export class PrevFinalPosSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([PositionComponent, PrevFinalPosComponent]);
        for (const entityId of entities) {
            const posComponent = world.getComponent(entityId, PositionComponent);
            const prevFinalPosComponent = world.getComponent(entityId,PrevFinalPosComponent);
            prevFinalPosComponent.pos.set(posComponent.pos);
        }
    }
}


export class AngularMovementSystem {
    runInPause = false;
    update(world, dt) {
        const entities = world.query([OrientationComponent, AngularVelocityComponent]);
        for (const entityId of entities) {
            const orientation = world.getComponent(entityId, OrientationComponent);
            const angularVel = world.getComponent(entityId, AngularVelocityComponent);

            orientation.angle += angularVel.angularVelocity * dt;
        }
    }
}

export class PBDBallBallCollisions {
  runInPause = false;
  update(world, dt) {
    const ballEntities = world.query([BallTagComponent, PositionComponent, RadiusComponent, MassComponent]);
    for (let i = 0; i < ballEntities.length; i++) {
      for (let j = i + 1; j < ballEntities.length; j++) {
        const e1 = ballEntities[i];
        const e2 = ballEntities[j];

        const p1Comp = world.getComponent(e1, PositionComponent);
        const r1 = world.getComponent(e1, RadiusComponent).radius;
        const m1Comp = world.getComponent(e1, MassComponent);

        const p2Comp = world.getComponent(e2, PositionComponent);
        const r2 = world.getComponent(e2, RadiusComponent).radius;
        const m2Comp = world.getComponent(e2, MassComponent);

        const p1 = p1Comp.pos;
        const m1 = m1Comp ? m1Comp.mass : 0.0;
        const p2 = p2Comp.pos;
        const m2 = m2Comp ? m2Comp.mass : 0.0;

        const dir = new Vector2().subtractVectors(p2, p1);
        const dSq = dir.lengthSq();
        const rSum = r1 + r2;

        if (dSq == 0.0 || dSq > rSum * rSum) continue;

        const d = Math.sqrt(dSq);
        dir.scale(1.0 / d); // Normalize

        // Resolve penetration
        const penetration = rSum - d;
        const invMass1 = (m1 > 0) ? 1.0 / m1 : 0.0;
        const invMass2 = (m2 > 0) ? 1.0 / m2 : 0.0;
        const totalInvMass = invMass1 + invMass2;

        if (totalInvMass <= 1e-9) continue;

        const corr = dir.clone().scale(penetration / totalInvMass);
        p1.add(corr, -invMass1);
        p2.add(corr, invMass2);
      }
    }
  }
}

export class PBDBallObstacleCollisions {
  runInPause = false;
  update(world, dt) {
    const ballEntities = world.query([BallTagComponent, PositionComponent, RadiusComponent]);
    const obstacleEntities = world.query([ObstacleTagComponent, PositionComponent, RadiusComponent, ObstaclePushComponent]);

    let contacts = world.getResource('ball_obstacle_contacts');
    if (!contacts) {
        contacts = [];
        world.setResource('ball_obstacle_contacts', contacts);
    }
    contacts.length = 0; // Clear existing contacts

    for (const ballId of ballEntities) {
      const p1 = world.getComponent(ballId, PositionComponent).pos;
      const r1 = world.getComponent(ballId, RadiusComponent).radius;

      for (const obsId of obstacleEntities) {
        const p2 = world.getComponent(obsId, PositionComponent).pos;
        const r2 = world.getComponent(obsId, RadiusComponent).radius;

        const dir = new Vector2().subtractVectors(p1, p2);
        const dSq = dir.lengthSq();
        const rSum = r1 + r2;

        if (dSq == 0.0 || dSq > rSum * rSum) continue;

        const d = Math.sqrt(dSq);
        dir.scale(1.0 / d); // Normalize

        // Store contact info for the velocity-based bump system
        contacts.push({ ball_id: ballId, obs_id: obsId, direction: dir.clone() });

        // Resolve penetration
        const corr = rSum - d;
        p1.add(dir, corr);

        // Velocity resolution is now handled by BallObstacleBumpSystem
        const grabbed = world.getResource('grabbedBall');
        if (ballId !== grabbed) {
          world.addComponent(ballId, new ScoredTagComponent());
        }
      }
    }
  }
}

export class InputReplaySystem {
  runInPause = false;
  constructor(inputLog, inputSystem) {
    this.inputLog = inputLog;
    this.currentIndex = 0;
    this.frame = 0;
    this.inputSystem = inputSystem;
  }
  update(world, dt) {
    if (this.inputLog.length > 0) {
      if (this.inputSystem.frame === this.inputLog[0].frame) {
        const frame = this.inputLog.shift();
        this.inputSystem.clicks = frame.clicks.slice();
        this.inputSystem.releases = frame.releases.slice();
      }
    }
  }
}

export class PBDBallFlipperCollisions {
  _getFlipperTip(flipperPos, flipperState) {
    const angle = flipperState.restAngle + flipperState.sign * flipperState.rotation;
    const dir = new Vector2(Math.cos(angle), Math.sin(angle));
    return flipperPos.clone().add(dir, flipperState.length);
  }

  update(world, dt) {
    const ballEntities = world.query([BallTagComponent, PositionComponent, RadiusComponent, MassComponent]);
    const flipperEntities = world.query([FlipperTagComponent, PositionComponent, RadiusComponent, FlipperStateComponent]);

    let contacts = world.getResource('ball_flipper_contacts');
    if (!contacts) {
        contacts = [];
        world.setResource('ball_flipper_contacts', contacts);
    }
    contacts.length = 0;

    for (const ballId of ballEntities) {
      const p1Comp = world.getComponent(ballId, PositionComponent);
      const p1 = p1Comp.pos;
      const r1 = world.getComponent(ballId, RadiusComponent).radius;
      const massComp = world.getComponent(ballId, MassComponent);
      const invMass = (massComp && massComp.mass > 0) ? 1.0 / massComp.mass : 0.0;

      for (const flipId of flipperEntities) {
        const fp = world.getComponent(flipId, PositionComponent).pos;
        const fr = world.getComponent(flipId, RadiusComponent).radius;
        const fs = world.getComponent(flipId, FlipperStateComponent);

        const tip = this._getFlipperTip(fp, fs);
        const closest = closestPointOnSegment(p1, fp, tip);

        const dir = new Vector2().subtractVectors(p1, closest);
        const dSq = dir.lengthSq();
        const rSum = r1 + fr;

        if (dSq == 0.0 || dSq > rSum * rSum) continue;

        const d = Math.sqrt(dSq);
        dir.scale(1.0 / d);

        const corr = rSum - d;
        if (invMass > 0) {
            p1.add(dir, corr);
        }

        let delta_lambda = 0;
        if (invMass > 0) {
            const w_inv = invMass;
            delta_lambda = corr / w_inv;
        }

        contacts.push({
            'ball_id': ballId,
            'flip_id': flipId,
            'normal': dir.clone(),
            'contact_point_on_flipper': closest.clone(),
            'delta_lambda': delta_lambda
        });
      }
    }
  }
}

export class PBDBallBorderCollisions {
  update(world, dt) {
    const ballEntities = world.query([BallTagComponent, PositionComponent, RadiusComponent, MassComponent]);
    const borderEntities = world.query([BorderComponent]);
    if (borderEntities.length === 0) return;

    const borderId = borderEntities[0];
    const borderComp = world.getComponent(borderId, BorderComponent);
    const borderPoints = borderComp.points;

    let contacts = world.getResource('ball_border_contacts');
    if (!contacts) {
        contacts = [];
        world.setResource('ball_border_contacts', contacts);
    }
    contacts.length = 0;

    if (borderPoints.length >= 2) {
        for (const ballId of ballEntities) {
            const p1Comp = world.getComponent(ballId, PositionComponent);
            const p1 = p1Comp.pos;
            const r1 = world.getComponent(ballId, RadiusComponent).radius;
            const massComp = world.getComponent(ballId, MassComponent);
            const invMass = (massComp && massComp.mass > 0) ? 1.0 / massComp.mass : 0.0;

            let minDistSq = Infinity;
            let closestSegPoint = new Vector2();
            let edgeStart = null;
            let edgeEnd = null;

            for (let i = 0; i < borderPoints.length; i++) {
                const a = borderPoints[i];
                const b = borderPoints[(i + 1) % borderPoints.length];
                const closestPtOnSeg = closestPointOnSegment(p1, a, b);
                const distSq = p1.distanceToSq(closestPtOnSeg);

                if (distSq < minDistSq) {
                    minDistSq = distSq;
                    closestSegPoint.set(closestPtOnSeg);
                    edgeStart = a;
                    edgeEnd = b;
                }
            }

            if (minDistSq > r1 * r1) continue;

            const ballToClosest = new Vector2().subtractVectors(p1, closestSegPoint);
            const edgeVec = new Vector2().subtractVectors(edgeEnd, edgeStart);
            const normal = new Vector2(-edgeVec.y, edgeVec.x).normalize();

            let collisionNormal;
            if (ballToClosest.lengthSq() < 1e-9) {
                collisionNormal = normal.clone();
            } else {
                collisionNormal = ballToClosest.clone().normalize();
            }

            if (ballToClosest.dot(normal) < 0) {
                collisionNormal = normal.clone();
            }

            const dist = Math.sqrt(minDistSq);
            const penetration = r1 - dist;
            let delta_lambda = 0;
            if (penetration > 0) {
                if (invMass > 0) {
                    p1.add(collisionNormal, penetration);
                    const w_inv = invMass;
                    delta_lambda = penetration / w_inv;
                }
            }

            contacts.push({
                'ball_id': ballId,
                'normal': collisionNormal.clone(),
                'delta_lambda': delta_lambda
            });
        }
    }
  }
}

File: python/common_systems.py
import math
import numpy as np
from . import vector2 as v2_helpers
from .ecs import (
    PositionComponent,
    VelocityComponent,
    GravityAffectedComponent,
    BallTagComponent,
    RadiusComponent,
    MassComponent,
    RestitutionComponent,
    PrevFinalPosComponent,
    OrientationComponent,
    MomentOfInertiaComponent,
    AngularVelocityComponent,
    CoefficientOfFrictionComponent,
    ObstacleTagComponent,
    ObstaclePushComponent,
    ScoredTagComponent,
    DistanceConstraintComponent,
    PrevFinalOrientationComponent
)


class PrevFinalOrientationSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([OrientationComponent, PrevFinalOrientationComponent])
        for entity_id in entities:
            orientation_comp = world.get_component(entity_id, OrientationComponent)
            prev_final_orientation_comp = world.get_component(entity_id, PrevFinalOrientationComponent)
            prev_final_orientation_comp.angle = orientation_comp.angle

class PBDAngularVelocityUpdateSystem:
    run_in_pause = False

    def _normalize_angle(self, angle): # Helper to find the shortest angle difference
        while angle > math.pi:
            angle -= 2 * math.pi
        while angle < -math.pi:
            angle += 2 * math.pi
        return angle

    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        entities = world.query([
            OrientationComponent,
            AngularVelocityComponent,
            PrevFinalOrientationComponent,
            MomentOfInertiaComponent
        ])
        epsilon = 1e-9

        if dt <= epsilon:
            return

        for entity_id in entities:
            if entity_id == grabbed:
                continue

            moi_comp = world.get_component(entity_id, MomentOfInertiaComponent)
            if moi_comp and moi_comp.invInertia <= 0: # Skip static or non-rotational objects
                continue

            orientation_comp = world.get_component(entity_id, OrientationComponent)
            angular_vel_comp = world.get_component(entity_id, AngularVelocityComponent)
            prev_final_orientation_comp = world.get_component(entity_id, PrevFinalOrientationComponent)

            current_angle = orientation_comp.angle
            prev_angle = prev_final_orientation_comp.angle

            delta_angle = self._normalize_angle(current_angle - prev_angle)

            angular_vel_comp.angularVelocity = delta_angle / dt

class PBDVelocityUpdateSystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        entities = world.query([PositionComponent, VelocityComponent, PrevFinalPosComponent, MassComponent])
        epsilon = 1e-9

        if dt <= epsilon: # Avoid division by zero or very small dt
            return

        for entity_id in entities:
            if entity_id == grabbed: # Don't update velocity of grabbed ball this way
                continue

            mass_comp = world.get_component(entity_id, MassComponent)
            # Only update velocities for dynamic objects (mass > 0)
            if mass_comp and mass_comp.mass <= 0:
                continue

            pos_comp = world.get_component(entity_id, PositionComponent)
            vel_comp = world.get_component(entity_id, VelocityComponent)
            prev_final_pos_comp = world.get_component(entity_id, PrevFinalPosComponent)

            # v = (x_new - x_old) / dt
            vel_comp.vel = (pos_comp.pos - prev_final_pos_comp.pos) / dt

class GravitySystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        gravity = world.get_resource('gravity')
        if gravity is None:
            return

        entities = world.query([VelocityComponent, GravityAffectedComponent])
        for entity_id in entities:
            if entity_id == grabbed:
                continue
            vel_comp = world.get_component(entity_id, VelocityComponent)
            vel_comp.vel += gravity * dt

class XPBDDistanceConstraintSystem:
    run_in_pause = False
    def update(self, world, dt):
        constraint_entities = world.query([DistanceConstraintComponent])
        epsilon = 1e-9 # Small value to avoid division by zero

        for entity_id in constraint_entities:
            constraint = world.get_component(entity_id, DistanceConstraintComponent)

            entity_a = constraint.entityA
            entity_b = constraint.entityB

            pos_a_comp = world.get_component(entity_a, PositionComponent)
            pos_b_comp = world.get_component(entity_b, PositionComponent)

            if not pos_a_comp or not pos_b_comp:
                continue

            p_a = pos_a_comp.pos
            p_b = pos_b_comp.pos

            mass_a_comp = world.get_component(entity_a, MassComponent)
            inv_mass_a = 1.0 / mass_a_comp.mass if mass_a_comp and mass_a_comp.mass > 0 else 0.0
            moi_a_comp = world.get_component(entity_a, MomentOfInertiaComponent)
            inv_inertia_a = moi_a_comp.invInertia if moi_a_comp else 0.0

            mass_b_comp = world.get_component(entity_b, MassComponent)
            inv_mass_b = 1.0 / mass_b_comp.mass if mass_b_comp and mass_b_comp.mass > 0 else 0.0
            moi_b_comp = world.get_component(entity_b, MomentOfInertiaComponent)
            inv_inertia_b = moi_b_comp.invInertia if moi_b_comp else 0.0

            if inv_mass_a + inv_mass_b + inv_inertia_a + inv_inertia_b <= epsilon:
                continue

            diff = p_b - p_a
            current_length = np.linalg.norm(diff)
            if current_length <= epsilon:
                continue

            direction = diff / current_length

            C = current_length - constraint.restLength

            alpha_tilde = constraint.compliance / (dt * dt)

            w_sum = inv_mass_a + inv_mass_b

            denominator = w_sum + alpha_tilde
            if denominator <= epsilon:
                continue
            
            # Note: `constraint.lambda` from JS is renamed to `constraint.lambda_val`
            # to avoid conflict with Python's `lambda` keyword.
            # The `DistanceConstraintComponent` dataclass should reflect this.
            delta_lambda = (-C - alpha_tilde * constraint.lambda_val) / denominator
            constraint.lambda_val += delta_lambda

            correction = direction * delta_lambda

            if inv_mass_a > 0:
                dp_a = correction * -inv_mass_a
                p_a += dp_a
            if inv_mass_b > 0:
                dp_b = correction * inv_mass_b
                p_b += dp_b

class MovementSystem:
    run_in_pause = False
    def update(self, world, dt):
        grabbed = world.get_resource('grabbedBall')
        linear_entities = world.query([PositionComponent, VelocityComponent])
        for entity_id in linear_entities:
            if entity_id == grabbed:
                continue
            pos_comp = world.get_component(entity_id, PositionComponent)
            vel_comp = world.get_component(entity_id, VelocityComponent)
            pos_comp.pos += vel_comp.vel * dt

class PrevFinalPosSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([PositionComponent, PrevFinalPosComponent])
        for entity_id in entities:
            pos_component = world.get_component(entity_id, PositionComponent)
            prev_final_pos_component = world.get_component(entity_id, PrevFinalPosComponent)
            prev_final_pos_component.pos[:] = pos_component.pos

class AngularMovementSystem:
    run_in_pause = False
    def update(self, world, dt):
        entities = world.query([OrientationComponent, AngularVelocityComponent])
        for entity_id in entities:
            orientation = world.get_component(entity_id, OrientationComponent)
            angular_vel = world.get_component(entity_id, AngularVelocityComponent)
            orientation.angle += angular_vel.angularVelocity * dt

class PBDBallBallCollisions:
    run_in_pause = False
    def update(self, world, dt):
        ball_entities = world.query([BallTagComponent, PositionComponent, VelocityComponent, RadiusComponent, MassComponent, RestitutionComponent])
        for i in range(len(ball_entities)):
            for j in range(i + 1, len(ball_entities)):
                e1 = ball_entities[i]
                e2 = ball_entities[j]

                p1 = world.get_component(e1, PositionComponent).pos
                v1 = world.get_component(e1, VelocityComponent).vel
                r1 = world.get_component(e1, RadiusComponent).radius
                m1 = world.get_component(e1, MassComponent).mass
                res1 = world.get_component(e1, RestitutionComponent).restitution

                p2 = world.get_component(e2, PositionComponent).pos
                v2 = world.get_component(e2, VelocityComponent).vel
                r2 = world.get_component(e2, RadiusComponent).radius
                m2 = world.get_component(e2, MassComponent).mass
                res2 = world.get_component(e2, RestitutionComponent).restitution

                restitution = min(res1, res2)
                direction = p2 - p1
                d_sq = np.dot(direction, direction)
                r_sum = r1 + r2

                if d_sq == 0.0 or d_sq > r_sum * r_sum:
                    continue

                d = math.sqrt(d_sq)
                direction /= d # Normalize

                # Resolve penetration
                corr = (r_sum - d) / 2.0
                p1 += direction * -corr
                p2 += direction * corr

                # Resolve velocity
                vel1_dot = np.dot(v1, direction)
                vel2_dot = np.dot(v2, direction)

                new_v1_dot = (m1 * vel1_dot + m2 * vel2_dot - m2 * (vel1_dot - vel2_dot) * restitution) / (m1 + m2)
                new_v2_dot = (m1 * vel1_dot + m2 * vel2_dot - m1 * (vel2_dot - vel1_dot) * restitution) / (m1 + m2)

                v1 += direction * (new_v1_dot - vel1_dot)
                v2 += direction * (new_v2_dot - vel2_dot)

class PBDBallObstacleCollisions:
    def update(self, world, dt):
        ball_entities = world.query([BallTagComponent, PositionComponent, VelocityComponent, RadiusComponent, MassComponent])
        obstacle_entities = world.query([ObstacleTagComponent, PositionComponent, RadiusComponent, ObstaclePushComponent])

        for ball_id in ball_entities:
            p1 = world.get_component(ball_id, PositionComponent).pos
            v1 = world.get_component(ball_id, VelocityComponent).vel
            r1 = world.get_component(ball_id, RadiusComponent).radius
            ball_mass_comp = world.get_component(ball_id, MassComponent)
            ball_ang_vel_comp = world.get_component(ball_id, AngularVelocityComponent)
            ball_moi_comp = world.get_component(ball_id, MomentOfInertiaComponent)

            for obs_id in obstacle_entities:
                p2 = world.get_component(obs_id, PositionComponent).pos
                r2 = world.get_component(obs_id, RadiusComponent).radius
                push_vel = world.get_component(obs_id, ObstaclePushComponent).pushVel

                direction = p1 - p2
                d_sq = np.dot(direction, direction)
                r_sum = r1 + r2

                if d_sq == 0.0 or d_sq > r_sum * r_sum:
                    continue

                d = math.sqrt(d_sq)
                direction /= d # Normalize

                # Resolve penetration
                corr = r_sum - d
                p1 += direction * corr

                # Resolve velocity & rotation with friction and obstacle rotation
                obs_ang_vel_comp = world.get_component(obs_id, AngularVelocityComponent)
                obs_moi_comp = world.get_component(obs_id, MomentOfInertiaComponent)
                obs_friction_comp = world.get_component(obs_id, CoefficientOfFrictionComponent)

                omega_obs = obs_ang_vel_comp.angularVelocity if obs_ang_vel_comp else 0.0
                mu = obs_friction_comp.mu if obs_friction_comp else 0.0

                tangent = np.array([-direction[1], direction[0], direction[2]])

                v_surf_obs_tangential_comp = 0
                if omega_obs != 0:
                    v_surf_obs_tangential_comp = (-omega_obs * direction[1] * r2) * tangent[0] + (omega_obs * direction[0] * r2) * tangent[1]

                s_sign = 0
                if v_surf_obs_tangential_comp != 0 and mu != 0:
                    s_sign = math.copysign(1, v_surf_obs_tangential_comp)
                
                effective_push_dir = direction.copy()
                if s_sign != 0:
                    effective_push_dir += tangent * (mu * s_sign)
                v2_helpers.normalize_inplace(effective_push_dir)

                v1 += effective_push_dir * push_vel

                if mu > 0 and ball_mass_comp:
                    delta_v_ball_tangential_actual_scalar_comp = np.dot(effective_push_dir, tangent) * push_vel

                    if abs(delta_v_ball_tangential_actual_scalar_comp) > 1e-9:
                        j_t_on_ball_vec = tangent * (delta_v_ball_tangential_actual_scalar_comp * ball_mass_comp.mass)

                        if ball_ang_vel_comp and ball_moi_comp and ball_moi_comp.invInertia > 0:
                            r_contact_ball = direction * -r1
                            delta_l_ball = r_contact_ball[0] * j_t_on_ball_vec[1] - r_contact_ball[1] * j_t_on_ball_vec[0]
                            ball_ang_vel_comp.angularVelocity += delta_l_ball * ball_moi_comp.invInertia

                        if obs_ang_vel_comp and obs_moi_comp and obs_moi_comp.invInertia > 0:
                            j_t_on_obs_vec = j_t_on_ball_vec * -1
                            r_contact_obs = direction * r2
                            delta_l_obs = r_contact_obs[0] * j_t_on_obs_vec[1] - r_contact_obs[1] * j_t_on_obs_vec[0]
                            obs_ang_vel_comp.angularVelocity += delta_l_obs * obs_moi_comp.invInertia

                grabbed = world.get_resource('grabbedBall')
                if ball_id != grabbed:
                    world.add_component(ball_id, ScoredTagComponent())

class InputReplaySystem:
    run_in_pause = False
    def __init__(self, input_log, input_system):
        self.input_log = input_log
        self.current_index = 0
        self.frame = 0
        self.input_system = input_system

    def update(self, world, dt):
        if len(self.input_log) > 0:
            # Assuming input_log is a list of dicts with 'frame', 'clicks', 'releases'
            if self.input_system.frame == self.input_log[0]['frame']:
                frame_data = self.input_log.pop(0)
                self.input_system.clicks = frame_data['clicks'][:]
                self.input_system.releases = frame_data['releases'][:]
